crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -48,127 +225,284 @@ where
+                    .iter()
+                    .any(|fut| fut.headers.iter().any(|h| h.number == num))
+                {
+                    continue
+                }
+
+                if range.is_empty() {
+                    request_range.start = num;
+                } else if request_range.end + 1 == num {
+                    request_range.end = num;
+                } else {
+                    // Dispatch contiguous request.
+                    self.in_progress_queue.push(
+                        BodiesRequestFuture::new(
+                            Arc::clone(&self.client),
+                            Arc::clone(&self.consensus),
                         )
-                })
-                .buffered(self.concurrency)
-                .map_ok(|blocks| stream::iter(blocks.into_iter()).map(Ok))
-                .try_flatten(),
-        )
+                        .with_headers(
+                            self.query_headers(request_range.start, request_range.count())?,
+                        ),
+                    );
+                    // Clear the current request range
+                    request_range = Range::default();
+                }
+            }
+        }
+
+        // Put buffered bodies back into the buffer.
+        let mut bodies = bodies.into_iter().peekable();
+        while bodies.peek().is_some() {
+            let mut contigious_batch = Vec::from([bodies.next().unwrap().0]);
+            while bodies
+                .peek()
+                .map(|b| b.block_number() == contigious_batch.last().unwrap().block_number() + 1)
+                .unwrap_or_default()
+            {
+                contigious_batch.push(bodies.next().unwrap().0);
+            }
+
+            self.buffered_responses.push(OrderedBodiesResponse(contigious_batch));
+        }
+
+        self.header_range = range;
+        self.latest_queued_block_number = None;
+        tracing::trace!(target: "downloaders::bodies", range = ?self.header_range, "New header range set");
+        Ok(())
     }
 }
 
-impl<Client, Consensus> ConcurrentDownloader<Client, Consensus>
+impl<B, C, DB> Stream for ConcurrentDownloader<B, C, DB>
 where
-    Client: BodiesClient,
-    Consensus: ConsensusTrait,
+    B: BodiesClient + 'static,
+    C: ConsensusTrait + 'static,
+    DB: Database,
 {
-    /// Create a new concurrent downloader instance.
-    pub fn new(client: Arc<Client>, consensus: Arc<Consensus>) -> Self {
-        Self { client, consensus, retries: 3, batch_size: 100, concurrency: 5 }
-    }
+    type Item = Vec<BlockResponse>;
 
-    /// Set the number of blocks to fetch at the same time.
-    ///
-    /// Defaults to 100.
-    pub fn with_batch_size(mut self, batch_size: usize) -> Self {
-        self.batch_size = batch_size;
-        self
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if this.is_terminated() {
```

Comment:
we need to remove terminated because the stage can be reactivated again at some point, so it can just pick up a new range again

Reply:
i wonder if it's fine? here's one example:
1. execute stage for (0, 100)
2. calls downloader.set_range(0, 100)
3. stream yields 100 elements, stage completes
4. new stage is called for (100, 200)
5. calls downloader.set_range(100, 200) which dispatches the requests, and so `is_terminated()` will be false?

Reply:
will add a test to make this behavior clear

Reply:
added `can_download_new_range_after_termination`